File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        if(matrix[matrix.size()-1][matrix[matrix.size()-1].size()-1] < target) return false; 
        int left = 0, right = matrix.size()-1; 
        int pivot = 0; 
        while(left <= right){
            int mid = (right + left) / 2; 
            
            int leftMost = matrix[mid][0], rightMost = matrix[mid][matrix[mid].size()-1]; 
            if(leftMost <= target && rightMost >= target){
                pivot = mid;
                break; 
            } else if(leftMost > target){
                right = mid - 1; 
            } else{
                left = mid + 1; 
            }
        }
        left = 0;  
        right = matrix[0].size()-1; 
        while(left <= right){
            int mid = (right + left) / 2; 
            if(matrix[pivot][mid] == target){
                return true; 
            }

            if(matrix[pivot][mid] > target){
                right = mid - 1; 
            } else{
                left = mid + 1; 
            }
        }

        return false; 
    }
};

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int left = 0, right = matrix.length-1; 
        int targetIndex = 0; 
        while(left <= right){
            int mid = left + (right - left) / 2; 
            int leftMost = matrix[mid][0]; 
            int rightMost = matrix[mid][matrix[0].length-1]; 

            if(leftMost > target){
                right = mid - 1; 
            } else if(rightMost < target){
                left = mid + 1; 
            } else{
                targetIndex = mid;
                break; 
            }
        }

        left = 0; 
        right = matrix[0].length-1; 
        while(left <= right){
            int mid = left + (right - left) / 2; 

            if(matrix[targetIndex][mid] == target){
                return true; 
            }

            if(matrix[targetIndex][mid] > target){
                right = mid - 1; 
            } else{
                left = mid + 1; 
            }
        }


        return false; 
    }
}